File: src/mapInfo.ts
/**
 * Map detection. The carnage XML has no map field, but MCC leaves two
 * breadcrumbs in sibling folders under the same Temporary dir:
 *
 *  - UserContent\Halo3\Movie\asq_<scenario>_<crc>_<hexts>.mov — the temporary
 *    theater film, written seconds AFTER the game ends, with the base map's
 *    scenario name (truncated to 7 chars) in the filename.
 *  - UserContent\Halo3\Map\<hexts>.mvar — the map variant loaded for the game,
 *    written at game START; the variant's display name (e.g. "MLG CStruct
 *    TS8") is the first UTF-16BE string in the blob.
 *
 * Both are best-effort: films rotate quickly and built-in variants may not
 * write an .mvar, so either field can come back undefined.
 */

import { readdir, readFile, stat } from "node:fs/promises";
import { join } from "node:path";

export interface MapInfo {
  mapName?: string;
  mapVariant?: string;
}

/** Halo 3 scenario stems as MCC truncates them in film filenames (7 chars). */
const MAP_NAMES: Record<string, string> = {
  armory: "Rat's Nest",
  bunkerw: "Standoff",
  chill: "Narrows",
  chillou: "Cold Storage",
  constru: "Construct",
  cyberdy: "The Pit",
  deadloc: "High Ground",
  descent: "Assembly",
  docks: "Longshore",
  fortres: "Citadel",
  ghostto: "Ghost Town",
  guardia: "Guardian",
  isolati: "Isolation",
  lockout: "Blackout",
  midship: "Heretic",
  riverwo: "Valhalla",
  salvati: "Epitaph",
  sandbox: "Sandbox",
  shrine: "Sandtrap",
  sidewin: "Avalanche",
  snowbou: "Snowbound",
  spaceca: "Orbital",
  warehou: "Foundry",
  zanziba: "Last Resort",
};

/** Film written within a minute before the report up to this long after it. */
const FILM_BEFORE_MS = 60_000;
const FILM_AFTER_MS = 5 * 60_000;
/** An .mvar older than this is assumed stale (different variant replayed). */
const MVAR_MAX_AGE_MS = 4 * 60 * 60_000;
const POLL_MS = 3_000;

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

/** mtimeMs of every file with `ext` in `dir` ({} on a missing dir). */
async function mtimes(dir: string, ext: string): Promise<Map<string, number>> {
  const out = new Map<string, number>();
  let names: string[];
  try {
    names = await readdir(dir);
  } catch {
    return out;
  }
  for (const n of names) {
    if (!n.toLowerCase().endsWith(ext)) continue;
    const m = await stat(join(dir, n)).then((s) => s.mtimeMs).catch(() => 0);
    if (m) out.set(n, m);
  }
  return out;
}

/** Base map from the film closest in time to the report, or undefined. */
async function filmMapName(movieDir: string, playedAtMs: number): Promise<string | undefined> {
  const films = await mtimes(movieDir, ".mov");
  let best: string | undefined;
  let bestDist = Infinity;
  for (const [name, m] of films) {
    if (m < playedAtMs - FILM_BEFORE_MS || m > playedAtMs + FILM_AFTER_MS) continue;
    const dist = Math.abs(m - playedAtMs);
    if (dist < bestDist) {
      best = name;
      bestDist = dist;
    }
  }
  const stem = best?.match(/^asq_([a-z0-9]+)_/i)?.[1].toLowerCase();
  if (!stem) return undefined;
  return MAP_NAMES[stem] ?? stem[0].toUpperCase() + stem.slice(1);
}

/** Variant name from the newest .mvar written before (or with) the report. */
async function mvarVariant(mapDir: string, playedAtMs: number): Promise<string | undefined> {
  const mvars = await mtimes(mapDir, ".mvar");
  let best: string | undefined;
  let bestM = -Infinity;
  for (const [name, m] of mvars) {
    if (m > playedAtMs + FILM_BEFORE_MS || m < playedAtMs - MVAR_MAX_AGE_MS) continue;
    if (m > bestM) {
      best = name;
      bestM = m;
    }
  }
  if (!best) return undefined;
  try {
    const buf = await readFile(join(mapDir, best));
    return firstUtf16BeString(buf);
  } catch {
    return undefined;
  }
}

/** First printable UTF-16BE run of 4+ chars — the variant's display name. */
function firstUtf16BeString(buf: Buffer): string | undefined {
  let run = "";
  for (let i = 0; i + 1 < buf.length; i += 2) {
    const c = (buf[i] << 8) | buf[i + 1];
    if (c >= 0x20 && c < 0x7f) {
      run += String.fromCharCode(c);
    } else {
      if (run.length >= 4) return run.trim();
      run = "";
    }
  }
  return run.length >= 4 ? run.trim() : undefined;
}

/**
 * Find the map for a report by its mtime. The film lands ~7s after the XML,
 * so the live watcher passes `waitMs` to poll for it; backfill passes 0 and
 * takes whatever survived rotation.
 */
export async function findMapInfo(
  carnageDir: string,
  playedAtMs: number,
  waitMs = 0,
): Promise<MapInfo> {
  const movieDir = join(carnageDir, "UserContent", "Halo3", "Movie");
  const mapDir = join(carnageDir, "UserContent", "Halo3", "Map");

  const deadline = Date.now() + waitMs;
  let mapName = await filmMapName(movieDir, playedAtMs);
  while (!mapName && Date.now() < deadline) {
    await sleep(POLL_MS);
    mapName = await filmMapName(movieDir, playedAtMs);
  }

  return { mapName, mapVariant: await mvarVariant(mapDir, playedAtMs) };
}

File: cpp/src/mapinfo.cpp
#include "mapinfo.h"

#include <chrono>
#include <cmath>
#include <filesystem>
#include <thread>
#include <unordered_map>
#include <vector>

#include "carnage.h"  // fileMtimeMs
#include "util.h"

namespace fs = std::filesystem;

namespace {

// Halo 3 scenario stems as MCC truncates them in film filenames (7 chars).
const std::unordered_map<std::string, std::string> MAP_NAMES = {
    {"armory", "Rat's Nest"},   {"bunkerw", "Standoff"},   {"chill", "Narrows"},
    {"chillou", "Cold Storage"}, {"constru", "Construct"},  {"cyberdy", "The Pit"},
    {"deadloc", "High Ground"},  {"descent", "Assembly"},   {"docks", "Longshore"},
    {"fortres", "Citadel"},      {"ghostto", "Ghost Town"}, {"guardia", "Guardian"},
    {"isolati", "Isolation"},    {"lockout", "Blackout"},   {"midship", "Heretic"},
    {"riverwo", "Valhalla"},     {"salvati", "Epitaph"},    {"sandbox", "Sandbox"},
    {"shrine", "Sandtrap"},      {"sidewin", "Avalanche"},  {"snowbou", "Snowbound"},
    {"spaceca", "Orbital"},      {"warehou", "Foundry"},    {"zanziba", "Last Resort"},
};

// Film written within a minute before the report up to this long after it.
constexpr long long FILM_BEFORE_MS = 60'000;
constexpr long long FILM_AFTER_MS = 5 * 60'000;
// An .mvar older than this is assumed stale (different variant replayed).
constexpr long long MVAR_MAX_AGE_MS = 4LL * 60 * 60'000;
constexpr int POLL_MS = 3'000;

struct Entry {
    std::string name;  // filename only
    std::string path;
    long long mtimeMs = 0;
};

// Every file with `ext` in `dir` (empty on a missing dir).
std::vector<Entry> listWithExt(const std::string& dir, const std::string& ext) {
    std::vector<Entry> out;
    std::error_code ec;
    for (auto& e : fs::directory_iterator(dir, ec)) {
        if (!e.is_regular_file()) continue;
        std::string name = e.path().filename().string();
        std::string l = util::toLower(name);
        if (l.size() < ext.size() || l.compare(l.size() - ext.size(), ext.size(), ext) != 0)
            continue;
        std::string path = e.path().string();
        long long m = fileMtimeMs(path);
        if (m) out.push_back({name, path, m});
    }
    return out;
}

// Base map from the film closest in time to the report, or "".
std::string filmMapName(const std::string& movieDir, long long playedAtMs) {
    std::string best;
    long long bestDist = -1;
    for (const Entry& f : listWithExt(movieDir, ".mov")) {
        if (f.mtimeMs < playedAtMs - FILM_BEFORE_MS || f.mtimeMs > playedAtMs + FILM_AFTER_MS)
            continue;
        long long dist = std::llabs(f.mtimeMs - playedAtMs);
        if (bestDist < 0 || dist < bestDist) {
            best = f.name;
            bestDist = dist;
        }
    }
    // asq_<stem>_<crc>_<hexts>.mov
    std::string l = util::toLower(best);
    if (l.compare(0, 4, "asq_") != 0) return "";
    size_t end = l.find('_', 4);
    if (end == std::string::npos || end == 4) return "";
    std::string stem = l.substr(4, end - 4);
    auto it = MAP_NAMES.find(stem);
    if (it != MAP_NAMES.end()) return it->second;
    stem[0] = static_cast<char>(std::toupper(static_cast<unsigned char>(stem[0])));
    return stem;
}

// First printable UTF-16BE run of 4+ chars — the variant's display name.
std::string firstUtf16BeString(const std::string& buf) {
    std::string run;
    for (size_t i = 0; i + 1 < buf.size(); i += 2) {
        unsigned c = (static_cast<unsigned char>(buf[i]) << 8) |
                     static_cast<unsigned char>(buf[i + 1]);
        if (c >= 0x20 && c < 0x7f) {
            run += static_cast<char>(c);
        } else {
            if (run.size() >= 4) return util::trim(run);
            run.clear();
        }
    }
    return run.size() >= 4 ? util::trim(run) : "";
}

// Variant name from the newest .mvar written before (or with) the report.
std::string mvarVariant(const std::string& mapDir, long long playedAtMs) {
    std::string bestPath;
    long long bestM = -1;
    for (const Entry& f : listWithExt(mapDir, ".mvar")) {
        if (f.mtimeMs > playedAtMs + FILM_BEFORE_MS || f.mtimeMs < playedAtMs - MVAR_MAX_AGE_MS)
            continue;
        if (f.mtimeMs > bestM) {
            bestPath = f.path;
            bestM = f.mtimeMs;
        }
    }
    if (bestPath.empty()) return "";
    auto buf = util::readFile(bestPath);
    return buf ? firstUtf16BeString(*buf) : "";
}

}  // namespace

MapInfo findMapInfo(const std::string& carnageDir, long long playedAtMs, int waitMs) {
    fs::path base(carnageDir);
    std::string movieDir = (base / "UserContent" / "Halo3" / "Movie").string();
    std::string mapDir = (base / "UserContent" / "Halo3" / "Map").string();

    auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(waitMs);
    std::string mapName = filmMapName(movieDir, playedAtMs);
    while (mapName.empty() && std::chrono::steady_clock::now() < deadline) {
        std::this_thread::sleep_for(std::chrono::milliseconds(POLL_MS));
        mapName = filmMapName(movieDir, playedAtMs);
    }

    return {mapName, mvarVariant(mapDir, playedAtMs)};
}
